Yuhsuan/1306 velocity reference

YuHsuan-Hwang opened the pull request:
This PR closes #1306:
* added "Set Rest Frequency" to the right-click menu in the layer widget for opening the rest frequency dialog
* this feature is only enabled for 3D images with frequency/wavelength native spectral type
* after saving custom rest frequency, the converted velocity in the z profile and the spectrally matched channels will be updated

YuHsuan-Hwang committed:
add dialog

YuHsuan-Hwang committed:
adjust title; add unit dropdown

YuHsuan-Hwang committed:
save and update rest freq

YuHsuan-Hwang committed:
add NonIdealState

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1306_velocity_reference

YuHsuan-Hwang committed:
error handling; disable menu item for 2D images

YuHsuan-Hwang committed:
minor change

YuHsuan-Hwang committed:
disable for certain spectral types

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1306_velocity_reference

YuHsuan-Hwang committed:
adjust className

YuHsuan-Hwang committed:
fix disable by types

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1306_velocity_reference

kswang1029 reviewed (commented):
This works nicely. 👍 Image matching is also supported after a new rest frequency is defined.

Do you think for the PV image defined as OFFSET VS "FREQ" we can have the same capability? It would be great to support it too.


One minor improvement is we add the rest frequency defined in header in the dialog like " Header RESTFRQ = 1.152712018000E+11 / Rest Frequency (Hz) " somewhere so that it is easier to know if RESTFRQ has been overwritten or not. Or another way to let user know the current RESTFRQ has been overwritten.

YuHsuan-Hwang committed:
apply custom rest freq to pv image

YuHsuan-Hwang committed:
add header rest freq info in tooltip

YuHsuan-Hwang committed:
reformat

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1306_velocity_reference

YuHsuan-Hwang commented:
@kswang1029 
> Do you think for the PV image defined as OFFSET VS "FREQ" we can have the same capability? It would be great to support it too.

Yes, added this feature to PV images. Please test this part, thanks

> One minor improvement is we add the rest frequency defined in header in the dialog like " Header RESTFRQ = 1.152712018000E+11 / Rest Frequency (Hz) " somewhere so that it is easier to know if RESTFRQ has been overwritten or not. Or another way to let user know the current RESTFRQ has been overwritten.

Added the information in the tooltip. The reset button is disabled if RESTFRQ is not provided in the header

kswang1029 reviewed (commented):
Found two usability issues:
1. If we load an image cube then immediately set a new rest frequency, the displayed velocity in the cursor info is not updated accordingly. We will need to change to a new channel to make the new rest frequency effective in the velocity conversion. 
2. If we load an image cube then immediately set a new rest frequency, then we load the same image cube again in append mode and match the two image cubes, we will see initially the matched channel indices are identical. We should see they are different as now their rest frequencies are different. This is probably related to the first issue.

kswang1029 commented:
Another usability improvement: we should disable the "set rest frequency" menu when animation playback is on, just to simplify things to prevent unforeseen side effects.

kswang1029 commented:
> Found two usability issues:
> 
> 1. If we load an image cube then immediately set a new rest frequency, the displayed velocity in the cursor info is not updated accordingly. We will need to change to a new channel to make the new rest frequency effective in the velocity conversion.
> 2. If we load an image cube then immediately set a new rest frequency, then we load the same image cube again in append mode and match the two image cubes, we will see initially the matched channel indices are identical. We should see they are different as now their rest frequencies are different. This is probably related to the first issue.

my bad. the second issue is non issue. The behavior is correct.

YuHsuan-Hwang committed:
fix velocity string update

YuHsuan-Hwang committed:
diable save when animation is active

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1306_velocity_reference

YuHsuan-Hwang commented:
@kswang1029 
> 1. If we load an image cube then immediately set a new rest frequency, the displayed velocity in the cursor info is not updated accordingly. We will need to change to a new channel to make the new rest frequency effective in the velocity conversion.

Fixed this issue. The velocity string is automatically updated

> Another usability improvement: we should disable the "set rest frequency" menu when animation playback is on, just to simplify things to prevent unforeseen side effects.

I disabled the save button if the animation is active (in case the dialog is already open)

kswang1029 reviewed (approved):
nice. 

Later on we can also add a new tab in the image list setting dialog and let user configure rest frequency of loaded images when possible. The current context menu can be severed as a shortcut then.

veggiesaurus commented on src/components/Dialogs/RestFreqDialog/RestFreqDialogComponent.tsx:
```diff
@@ -0,0 +1,154 @@
+        }
+    };
+
+    private saveRestFreq = () => {
```
perhaps make this an `@action` to bundle the called actions together

YuHsuan-Hwang replied:
I'm not sure if it's necessary because only `hideRestFreqDialog()` is an `@action` function. or do we consider also add `@action` to `updateCustomRestFreq()`?

veggiesaurus replied:
yes, that should also be an `@action`, since it calls multiple actions. Functions that call multiple actions should also be actions, so that we reduce the number of reactions

YuHsuan-Hwang replied:
@veggiesaurus added `@action` to `saveRestFreq()` and `updateCustomRestFreq()`. The features work fine after the change

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1306_velocity_reference

YuHsuan-Hwang committed:
Merge branch 'dev' into yuhsuan/1306_velocity_reference

YuHsuan-Hwang committed:
fix conflicts and merge dev

YuHsuan-Hwang committed:
add action decorators

YuHsuan-Hwang committed:
reformat
